### ISSUE TITLE
1.4.1: "export 'parse' was not found in '@mliebelt/pgn-parser'"

### ISSUE THREAD

**Issue by kurko** [mliebelt/pgn-parser#136]
I upgraded from 1.1.1 to 1.4.1 and started getting this error. I looked around, tried the README code and can't make it work.

It seems like `parse` is indeed supposed to be exported. In `node_modules/@mliebelt/pgn-parser/lib/index.js`, I can see:

```js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.parse = exports.split = void 0;
var split_games_1 = require("./split-games");
Object.defineProperty(exports, "split", { enumerable: true, get: function () { return split_games_1.split; } });
var pgn_parser_1 = require("./pgn-parser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return pgn_parser_1.parse; } });
```

But somehow `parse` is not available...

**Comment by mliebelt** [comment]
Sorry to hear that, I forgot to update the `README.md` to give the right direction to require / import the parse function.

## TLDR

You should use the following line to import the parse functionality (stolen from https://github.com/mliebelt/PgnViewerJS/blob/main/modules/pgn-reader/src/pgn.js, first line)

`import { parse, split } from '@mliebelt/pgn-parser';`

If you don't need the  split functionality, just

`import { parse } from '@mliebelt/pgn-parser';`

should do it.

I will  update the `README.md` later, and see if other changes are necessary there as well. See the [Stackoverflow answer](https://stackoverflow.com/a/31783409/2603665) if you need require instead of import.

`var parse = require('@mliebelt/pgn-parse').parse;`

should be the require statement.

**Comment by leonardodosdosdev** [comment]
I am still experiencing this same problem for 1.4.1:

"export 'parse' (imported as 'parse') was not found in '@mliebelt/pgn-parser' (module has no exports)"

even after using:

import { parse } from '@mliebelt/pgn-parser'

or when using:

var parse = require('@mliebelt/pgn-parser').parse

parse is undefined

**Comment by mliebelt** [comment]
I have released a new version 1.4.2 this morning, because I had problems in using the pgn-parser in the pgn-reader, which was then used by pgn-viewer. For that version, I was able to use it stable by:

`let parse = require('@mliebelt/pgn-parser').parse`

I have problems in using the ES6 syntax, but have not yet the time to search for the error. If you have experiences in providing a typescript library in the 3 format node, AMD and UMD, you are welcome.

Please try that syntax for version 1.4.2 if that fixes your problem. Is there a repository (publicly) available that shows the usage? Just having that one line of code without any context, I am lost.

**Comment by mliebelt** [comment]
I just checked again the circumstance, when I see a similar error:

* Current main branch in https://github.com/mliebelt/PgnViewerJS
* Set the [line 2 instead of 1](https://github.com/mliebelt/PgnViewerJS/blob/b531433e9998a849a0b773704dc281a056d42da0/modules/pgn-reader/src/pgn.ts#L2) so use the import as you do.
* Call the `build` command for the module `pgn-reader`.
* On the build step `umd` I get a similar error.

````
> rollup lib-esm/pgn.js -c rollup.config.js -o lib/index.umd.js -f umd

lib-esm/pgn.js → lib/index.umd.js...
[!] Error: 'parse' is not exported by node_modules/@mliebelt/pgn-parser/lib/index.mjs, imported by lib-esm/pgn.js
https://rollupjs.org/guide/en/#error-name-is-not-exported-by-module
lib-esm/pgn.js (2:9)
````

Is that a similar context where you see the error? Please try out version 1.4.2, I cannot fix the error for 1.4.1.

**Comment by mliebelt** [comment]
After doing some more studies, experimenting with creating all 3 format with imports, I don't see a simple solution (at the moment). If someone wants to do some research themselves, here are some entry points:

* https://github.com/mliebelt/use-different-module-types/tree/main/modules/rollup-use-library This is an example of 3 different usage contexts from the same base. Looked promising, but I was not able to move that technique to pgn-parser. Will continue experimenting, but  ... well, after having invested >10 hours, I would like to do some different stuff.
* https://www.sensedeep.com/blog/posts/2021/how-to-create-single-source-npm-module.html this matches exactly what I try to do, but  at the moment, I have not the nerves to try that out. But this could be a valid, clean approach.

**Comment by georgehenderson** [comment]
Seeing the same issue above in the latest build. Importing as a commonjs format returns an empty module. It doesnt seem like these lines are working from the lib/index.js. I tried importing the ./pgn-parser directly but i think the issue may lie in the bundling of the _pgn-parser module with peggy

```
Object.defineProperty(exports, "split", { enumerable: true, get: function () { return split_games_1.split; } });
var pgn_parser_1 = require("./pgn-parser");
Object.defineProperty(exports, "parse", { enumerable: true, get: function () { return pgn_parser_1.parse; } });
```

**Comment by mliebelt** [comment]
@georgehenderson Thank you a lot for giving feedback. I have some questions about your comment:

* What do you mean by "latest build"? The one after the [latest commit](https://github.com/mliebelt/pgn-parser/commit/918891ff601f43405fc3c7e32f184cc9085b52e2)? Or the  build of [version 1.4.2](https://www.npmjs.com/package/@mliebelt/pgn-parser/v/1.4.2) on NPM?
* The file `lib/index.js` is available only in version 1.4.2 (and former ones). I have created an example on [use-different-module-types](https://github.com/mliebelt/use-different-module-types/tree/main/modules/rollup-use-library-pgn-parser42) that shows how to use the current version in a library. There both cases for import with require and import statement works. What is the difference here, why do they work?
* What is the problem then?
  * Do I have to file an issue for peggy, so that they fix something? What should be fixed there?
  * Could I fix manually the generated library, so that the import both in cjs and mjs generated code works again?
  * Should I generate 2 versions of the parser (== `_pgn-parser`), and use the right one in the generated code then? I tried that already, and have seen no effect on the result.
  * Should I fix something in the typescript compiler call then? And what and how to fix then? See the current state of the library (GitHub). The test cases run, the scripts in directory `doc` can be called, with the special case, that the file `read-sample-mjs.mjs` uses as import statement: `import { parse } from '../lib/index.umd.js'`. This is the one problem I see, where I have no fix at the  moment.

Thank you a lot for your feedback, I don't give up the hope to have a fix for that nasty problem. If you have ideas what to try out, please share it.

**Comment by georgehenderson** [comment]
I am using 1.4.2, the latest build on npm. 

I was able to get it to work by using the umd build like you recommended. So my code looks like this.

```javascript
const pgnParser = require('@mliebelt/pgn-parser/lib/index.umd');
const game = pgnParser.parse(pgn.replace(/\[FEN .*\]/, ''), {startRule: 'game'})
```

**Comment by mliebelt** [comment]
I will play a little bit with different versions, but at the end, I will provide only the UMD build, if that is working in all environments (both Node.js and Browser, in different versions). My goal is, that all kind of imports work nevertheless what the people are using. Providing CommonJS, AMD and UMD is perhaps a goal that no one is interested in.

**Comment by audiBookning** [comment]
I read this issue and remembered another one that i stumbled on and having a similar error and situation. So i will just throw here a random though.
Might this issue be related to the following?
[Typescript 3.8 - Type-Only Imports and Exports](https://devblogs.microsoft.com/typescript/announcing-typescript-3-8/#type-only-imports-exports )

**Comment by mliebelt** [comment]
@audiBookning Thank you for the hint, but I think the reason was, that I tried to much. Having the grammar compiled to UMD, and using that in pgn-parser.ts and transpiling to UMD again works. I will create a new version, that hopefully is backward compatible, and will work for old `require` and new `import` syntax in both node.js and the browser.
